package/MDAnalysis/core/selection.py

```diff
@@ -518,11 +518,19 @@ def __init__(self, parser, tokens):
         self.values = vals
 
     def apply(self, group):
-        mask = np.zeros(len(group), dtype=np.bool)
-        for val in self.values:
-            values = getattr(group, self.field)
-            mask |= [fnmatch.fnmatch(x, val) for x in values]
-        return group[mask].unique
+        # rather than work on group.names, cheat and look at the lookup table
+        nmattr = getattr(group.universe._topology, self.field)
+
+        matches = []  # list of passing indices
```

Comment:
The code in this function is duplicated several times. Would there be a way of factoring it out?

Reply:
The difference is here it's `group.ix` as we're getting an atom attribute, later it's `group.resindices`/`group.segindices` as it's resnames/segids.  We could have this as `getattr(group, self.attrlevel)`, where attrlevel = `ix | resindices | segindices`.  This is arguably a little less readable but more compact in terms of repetition.  Not fussed either way, what's easier to read/maintain?

Reply:
My concern is that if we get to fix or change anything here, the odds are that one of the copies will be left out leading to very confusing discrepancies. I would go for the `getattr` solution here. Though I am not doing the work, so I won't block the PR either way.

Reply:
Yeah that's a good point, I'll refactor so it breaks/fixes as one unit :)